pkg/store/bucket.go

```diff
@@ -1694,7 +1694,12 @@ func (s *BucketStore) Series(req *storepb.SeriesRequest, seriesSrv storepb.Store
 	tracing.DoInSpan(ctx, "bucket_store_merge_all", func(ctx context.Context) {
 		begin := time.Now()
 		set := NewResponseDeduplicator(NewProxyResponseLoserTree(respSets...))
+		i := 0
 		for set.Next() {
+			i++
+			if req.Limit > 0 && i > int(req.Limit) {
```

Comment:
I guess when we have postings we can decided to only fetch "Limit" many series?

Reply:
I think I would just limit expanded postings, iiuc then we would only fetch series for that limited set of postings which is probably enough. @yeya24 is that somewhat sensible?

As to the response set.. I think it should be eager in the store at least for Thanos since we need to resort on the server before returning a response generally.

Reply:
There might be an edge case where limiting the expanded postings may return less results than expected.

Assume `/api/v1/labels` is called with `Limit=2`. 

Assume store-gateways queries two blocks containing the following series.

```
Block A
---------
http_req_total{host="H1"}
http_req_total{host="H2"}
http_req_total{host="H3"}
http_req_total{env="E1"}

Block B
---------
http_req_total{host="H5"}
http_req_total{host="H6"}
http_req_total{host="H7"}
http_req_total{env="E2"}
```

If the expanded postings are limited 2, then only the following series will be returned by the `blockSeriesClient`.

```
Block A
---------
http_req_total{host="H1"}
http_req_total{host="H2"}
```

```
Block B
---------
http_req_total{host="H5"}
http_req_total{host="H6"}
```

The final results will only contain the `host` label. It'll not return the `env` label.


This edge case can also apply to LabelValues(). 
`Select()` is used in `/api/v1/labels` and `/api/v1/label/<name>/values` with matchers, this also applies there.

Is there a way to handle such cases @yeya24 @MichaHoffmann ?

Reply:
Apologies, I misunderstood the code. It makes sense to do this for `Series()` calls and not `LabelValues()` or `LabelNames()`. I'll update the PR with this change.